cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+            target_entities["tabs"].update(get_tabs_from_app(root))
+
+    return target_entities
+
+
+def clean_zip_file(zip_src: zipfile.ZipFile, target_entities: Dict[str, set]):
+    zip_dest = zipfile.ZipFile(io.BytesIO(), "w", zipfile.ZIP_DEFLATED)
+    for name in zip_src.namelist():
+        file = zip_src.open(name)
+        if any(
+            item.extension in name and name.startswith(item.folder_name)
+            for item in FILES_TO_BE_CLEANED
+        ):
+            root = ET.parse(file).getroot()
+            cleaned_root = CleanXML(root, target_entities)
+            tree = ET.ElementTree(cleaned_root)
+            cleaned_content = io.BytesIO()
+            tree.write(
+                cleaned_content,
+                pretty_print=True,
+                xml_declaration=True,
+                encoding="utf-8",
+            )
+            zip_dest.writestr(name, cleaned_content.getvalue())
+        else:
+            zip_dest.writestr(name, file.read())
+
+    return zip_dest
+
+
+def strip_namespace(element: ET.Element):
+    for elem in element.iter():
+        if "}" in elem.tag:
+            elem.tag = elem.tag.split("}", 1)[1]
+    return element
+
+
+def fetch_permissionable_entity_names(
+    root: ET.Element, perm_entity: PermissionElementXPath, parent: bool = False
+):
+    entity_names = set()
+    for element in root.findall(perm_entity.permission_xpath):
+        entity_names.add(
+            perm_entity.return_parent(element)
+            if parent
+            else perm_entity.return_name(element)
+        )
+    return entity_names
+
+
+def CleanXML(root: ET.Element, target_entities: Dict[str, set]):
+    root = strip_namespace(root)
+    for key, perm_entities in FOLDER_PERM_DICT.items():
+        for perm_entity in perm_entities:
+            for element in root.findall(perm_entity.permission_xpath):
+                if perm_entity.return_name(element) not in target_entities.get(key, []):
+                    root.remove(element)
+    return root
+
+
+def entities_from_package(zf: zipfile.ZipFile, context: TaskContext, api_version: str):
+    package_xml = return_package_xml_from_zip(zf, api_version)
+    api = ApiRetrieveUnpackaged(
+        context, package_xml=package_xml, api_version=api_version
+    )
+    context.logger.info("Retrieving entities from package.xml")
+    retrieved_zf = api()
+    return get_target_entities_from_zip(retrieved_zf)
+
+
+def ret_sf(context: TaskContext, api_version: str):
+    sf = get_simple_salesforce_connection(
+        context.project_config,
+        context.org_config,
+        api_version=api_version,
+        base_url=None,
+    )
+    return sf
+
+
+def zip_clean_invalid_references(zf: zipfile.ZipFile, context: TaskContext):
+    # Set API version
+    api_version = context.org_config.latest_api_version
+
+    # Query and get entities
+    sf = ret_sf(context, api_version)
+    userPermissions, fields, tabs, objects = entities_from_query(sf)
```

Comment:
```suggestion
    user_permissions, fields, tabs, objects = entities_from_query(sf)
```

Takes a while to switch from Java's lowerCamelCase to Python's lower_snake_case :)

Reply:
yep will change that